docs/resources/server/recipes/custom.md

```diff
@@ -0,0 +1,756 @@
+                    var recipe = holder.value();
+                    inputStates.add(recipe.getInputState());
+                    inputItems.addAll(recipe.getInputItem().items());
+                });
+            
+            this.inputStates = Set.unmodifiableSet(inputStates);
+            this.inputItems = Set.unmodifiableSet(inputItems);
+        }
+    }
+
+    public void syncToClient(Stream<ServerPlayer> players) {
+        ClientboundRightClickBlockRecipesPayload payload =
+            new ClientboundRightClickBlockRecipesPayload(this.inputStates, this.inputItems);
+        players.forEach(player -> PacketDistributor.sendToPlayer(player, payload));
+    }
+
+    @Override
+    public Set<BlockState> inputStates() {
+        return this.inputStates;
+    }
+
+    @Override
+    public Set<Holder<Item>> inputItems() {
+        return this.inputItems;
+    }
+}
+
+// Client implementation to hold the inputs.
+public record ClientRightClickBlockRecipeInputs(
+    Set<BlockState> inputStates, Set<Holder<Item>> inputItems
+) implements RightClickBlockRecipeInputs {
+
+    public ClientRightClickBlockRecipeInputs(Set<BlockState> inputStates, Set<Holder<Item>> inputItems) {
+        this.inputStates = Set.unmodifiableSet(inputStates);
+        this.inputItems = Set.unmodifiableSet(inputItems);
+    }
+}
+
+// Handling the recipe instance depending on side.
+public class ServerRightClickBlockRecipes {
+
+    private static ServerRightClickBlockRecipeInputs inputs;
+
+    public static RightClickBlockRecipeInputs inputs() {
+        return ServerRightClickBlockRecipes.inputs;
+    }
+
+    // On the game event bus
+    @SubscribeEvent
+    public static void addListener(AddReloadListenerEvent event) {
+        // Register server reload listener
+        ServerRightClickBlockRecipes.inputs = new ServerRightClickBlockRecipeInputs(
+            event.getServerResources().getRecipeManager()
+        );
+        event.addListener(ServerRightClickBlockRecipes.inputs);
+    }
+
+    // On the game event bus
+    @SubscribeEvent
+    public static void datapackSync(OnDatapackSyncEvent event) {
+        // Send to client
+        ServerRightClickBlockRecipes.inputs.syncToClient(event.getRelevantPlayers());
+    }
+}
+public class ClientRightClickBlockRecipes {
+
+    private static ClientRightClickBlockRecipeInputs inputs;
+
+    public static RightClickBlockRecipeInputs inputs() {
+        return ClientRightClickBlockRecipes.inputs;
+    }
+
+    // Handling the sent packet
+    public static void handle(final ClientboundRightClickBlockRecipesPayload data, final IPayloadContext context) {
+        // Do something with the data, on the main thread
+        ClientRightClickBlockRecipes.inputs = new ClientRightClickBlockRecipeInputs(
+            data.inputStates(), data.inputItems()
+        );
+    }
+}
+
+public class RightClickBlockRecipes {
+    // Make proxy method to access properly
+    public static RightClickBlockRecipeInputs inputs() {
```

Comment:
This method needs to take in either the `Level#isClientSide` flag or a `Level` from which it retrieves that flag. Checking the physical distribution doesn't fix the issue of using the same cache on both threads in singleplayer.

Reply:
My brain for some reason remembered this incorrectly.